File: src/net/curl/curl.c
//go:build ignore

#include "curl.h"
#include <curl/curl.h>
#include <stdint.h>
#include <string.h>
#include "so/builtin/builtin.h"

bool CurlInit() {
  CURLcode res = curl_global_init(CURL_GLOBAL_DEFAULT);
  return res == CURLE_OK;
}
void CurlDeInit() {
  curl_global_cleanup();
}
// Closes the socket. Does not wait for all data to be sent.
void CurlCloseSocket(CURL* curl) {
  curl_easy_cleanup(curl);
}
so_String CurlErrorimpl(void* self){
  CURLcode code = (intptr_t)self;
  return so_str(curl_easy_strerror(code));
}
so_Error newCurlError(CURLcode code) {
  auto s = so_str(curl_easy_strerror(code));
  return (so_Error){
      .self = (void*)(intptr_t)code,
      .Error = CurlErrorimpl,
  };
}
// NOTE: hostname must be prefixed with "http://"
so_Error CurlCreateSocket(const char* hostname, CURL** curlRet) {
  CURL* curl = curl_easy_init();
  if (!curl) {
    return newCurlError(CURLE_FAILED_INIT);
  }

  curl_easy_setopt(curl, CURLOPT_URL, hostname);
  // raw TCP connection only
  curl_easy_setopt(curl, CURLOPT_CONNECT_ONLY, 1L);

  // connect to the server.
  CURLcode result = curl_easy_perform(curl);
  if (result != CURLE_OK) {
    return newCurlError(result);
  }
  *curlRet = curl;
  return (so_Error){};
}
// Returns the number of bytes read. Can be 0. -1 means error.
// This is non-blocking. It will not fill the buffer if there is no data.
so_R_int_err CurlReadFromSocket(CURL* curl, void* buffer, size_t buflen) {
  size_t n;
  CURLcode res = curl_easy_recv(curl, buffer, buflen, &n);
  if (res == CURLE_AGAIN)
    return (so_R_int_err){.val = 0, .err = NULL};
  if (res != CURLE_OK) {
    return (so_R_int_err){.val = -1, .err = newCurlError(res)};
  }
  return (so_R_int_err){.val = n, .err = NULL};
}

// This is non-blocking.
// Returns the number of bytes written. Can be 0. -1 means error.
so_R_int_err CurlWriteToSocket(CURL* curl, const void* buffer, size_t buflen) {
  size_t n;
  CURLcode res = curl_easy_send(curl, buffer, buflen, &n);
  if (res == CURLE_AGAIN)
    return (so_R_int_err){.val = 0, .err = NULL};
  if (res != CURLE_OK) {
    return (so_R_int_err){.val = -1, .err = newCurlError(res)};
  }
  return (so_R_int_err){.val = n, .err = NULL};
}

File: src/net/curl/curl.go
package curl

import (
	"mbc/sdl"
	"solod.dev/so/c"
	"solod.dev/so/time"
)

//so:include <curl/curl.h>

//so:extern CURL_GLOBAL_DEFAULT
const CURL_GLOBAL_DEFAULT = (1 << 0) | (1 << 1)

//so:extern CURLcode
type CURLcode int

//so:extern CURLE_AGAIN
const CURLE_AGAIN CURLcode = 1

//so:extern
func curl_global_init(int) CURLcode

//so:extern
func curl_easy_strerror(CURLcode) *c.ConstChar

//so:extern
func curl_easy_cleanup(*CURL)

//so:extern
func curl_easy_init() *CURL

//so:extern
func curl_easy_setopt(c *CURL, opt any, value any)

//so:extern
func curl_easy_perform(*CURL) CURLcode

//so:extern
type size_t int

//so:extern
func curl_easy_recv(curl *CURL, buf *byte, size int, n *size_t) CURLcode

//so:extern
func curl_easy_send(curl *CURL, buf *byte, size int, n *size_t) CURLcode

//so:extern CURLOPT_URL
const CURLOPT_URL = 0

//so:extern CURLOPT_CONNECT_ONLY
const CURLOPT_CONNECT_ONLY = 0

//so:extern CURLOPT_TCP_NODELAY
const CURLOPT_TCP_NODELAY = 0

//so:extern CURL
type CURL struct{}

type CurlError struct {
	code CURLcode
}

func (e *CurlError) Error() string {
	return c.String(curl_easy_strerror(e.code))
}

var _Error CurlError

// NOTE: hostname must be prefixed with "http://"
func CreateSocket(hostname string) (*CURL, error) {
	curl := curl_easy_init()
	curl_easy_setopt(curl, CURLOPT_URL, c.CString(hostname))
	curl_easy_setopt(curl, CURLOPT_TCP_NODELAY, c.Long(1))
	curl_easy_setopt(curl, CURLOPT_CONNECT_ONLY, c.Long(1))

	code := curl_easy_perform(curl)
	if code != 0 {
		_Error.code = code
		curl_easy_cleanup(curl)
		return nil, &_Error
	}
	return curl, nil
}

// Closes the socket. Does not wait for all data to be sent.
func CloseSocket(curl *CURL) {
	curl_easy_cleanup(curl)
}

// Returns the number of bytes read. Can be 0. -1 means error.
// This is non-blocking. It will not fill the buffer if there is no data.
func ReadFromSocket(curl *CURL, buffer *byte, size int) (int, error) {
	var n size_t
	code := curl_easy_recv(curl, buffer, size, &n)
	if code == CURLE_AGAIN {
		return int(n), nil
	} else {
	}
	if code != 0 {
		curl_easy_cleanup(curl)
		_Error.code = code
		return int(n), &_Error
	}
	return int(n), nil
}

// WriteToSocket blocks until all bytes are written or an error occurs.
// Returns the number of bytes written before an error, if any.
func WriteToSocket(curl *CURL, buffer *byte, size int) (int, error) {
	total := 0

	for total < size {
		var n size_t

		code := curl_easy_send(
			curl,
			buffer,
			size-total,
			&n,
		)

		total += int(n)

		switch code {
		case 0: // no error, keep writing.

		case CURLE_AGAIN:
			// Retry until writable.
			sdl.Delay(time.Millisecond)
			continue

		default:
			curl_easy_cleanup(curl)
			_Error.code = code
			return total, &_Error
		}
	}

	return total, nil
}

func init() {
	curl_global_init(CURL_GLOBAL_DEFAULT)
}
